docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+
+### Implementation Changes
+
+The "seams" in the P2P implementation between the higher level
+constructs (reactors), the routing layer (`Router`) and the lower
+level connection and peer management code make this operation
+relatively straight forward from an implementation perspective. The
+goal in this design is to minimize the impact to the reactors
+(potentially entirely,) and completely remove the lower level
+aspects (e.g. `Transport`, `Connection` and `PeerManager`) using the
+separation afforded by the `*Router` layer. The current state of the
+code makes these changes relatively surgical, and limited to a small
+number of methods:
+
+- `p2p.Router.OpenChannel` will still return a `Channel` structure
+  which will continue to serve as a pipe between the reactors and the
+  `Router`. The implementation will no longer need the queue
+  implementation and this operation will instead start goroutines that
+  are responsible for routing the messages from the channel to libp2p
+  fundamentals, replacing the current `p2p.Router.routeChannel`.
+
+- The current `p2p.Rotuer.dialPeers` and `p2p.Router.acceptPeers`,
+  which are responsible for establishing outbound and inbound
+  connections respectively, will be removed, along with
+  `p2p.Router.openConnection` and the libp2p connection manager will
+  be responsible for maintaining network connectivity.
+
+- The `p2p.Channel` interface should change to replace the golang
+  channels with a more functional interface for sending messages. The
+  new methods on this object, will take contexts to support safe
+  cancellation and return errors, and generally block rather than
+  return asynchronously. The `Out` channel through which
+  reactors send messages to Peers, will be replaced by a `Send`
+  method, and the Error channel will be replaced by an `Error`
+  method. 
+
+- Rectors will have access to an interface that will allow them to
+  access information from the current state of the Peer information
+  from libp2p. This will supplant the  `p2p.PeerUpdates`
+  subscription.
+
+### Upgrade and Compatibility
+
+Because the routers and all current P2P code is in the `internal`
+package and not part of the public API, the only changes to the public
+API surface area of the tendermint will be different configuration
+file options, replacing the current P2P options with options relevant
+to libp2p.
+
+However, it will not be possible to run a network with both networking
+stacks active at once, so the upgrade to the version of Tendermint
+will need to be coordinated between all nodes of the network.
+
+## Open Questions
+
+- What is the role of Protocol Labs in the implementation of libp2p in
+  tendermint, both during the initial implementation and on an ongoing
+  basis thereafter?
+
+- Should all P2P traffic for a given node be pushed to a single topic,
+  and we assume that a topic maps to a specific ChainID, or should
+  each reactor (or type of message) have its own topic?
```

Comment:
I'm not sure I understand the implications of the answer to this question. Previously, you mentioned topic-per-message as a possible strategy for obsoleting Router, but I thought that was mainly an implementation detail.

Is the question about whether the axis of variation for topics is (chanID x messageType) vs. just chainID vs. just messageType?

Do we have any information about how expensive it is to have a whole bunch of topics? It seems like libp2p wants topics to be lightweight and relatively inexpensive, but I don't have a good sense of the cost.

This ADR doesn't necessarily need to answer all that, but it may make sense to sketch out a few high-level criteria we can use for deciding once we know more.

Reply:
https://docs.libp2p.io/concepts/publish-subscribe/ is probably a decent place to start. Broadly: A "topic" is a label publishers and subscribers use to coordinate. But I don't know how how expensive that abstraction is in libp2p.

Reply:
Im a fan of per reactor topics. This would allow a form of service discovery in the network. Nodes that only run x reactor would only get subscribed to by peers looking for that reactor.

Reply:
I think the cost and scaling of the topic in libp2p is the like *core* question here, and I think we could go either way, if topics scale. 

As I've been thinking about it, I think I'm vaguely in the camp of one topic per logical consensus engine, but this might be a premature optimization for multi-chain, and because I think there's merit to retaining some amount of control over the message passing layer. 

This is definitely a thing we need PL feedback on.